TestConfigMapController test flake

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41412/pull-kubernetes-unit/17829/#k8siokubernetesfederationpkgfederation-controllerconfigmap-testconfigmapcontroller

```
panic: interface conversion: interface is nil, not *v1.ConfigMap
/usr/local/go/src/runtime/panic.go:500 +0x1ae
/usr/local/go/src/testing/testing.go:579 +0x474
/usr/local/go/src/runtime/panic.go:458 +0x271
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/configmap/configmap_controller_test.go:158 +0x74
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/configmap/configmap_controller_test.go:140 +0x1d8b
/usr/local/go/src/testing/testing.go:610 +0xca
/usr/local/go/src/testing/testing.go:646 +0x530
```

```
0xf2efbd	k8s.io/kubernetes/federation/pkg/federation-controller/util/test.RegisterFakeWatch.func1+0x14d	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/util/test/test_helper.go:155
0xa2799d	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).Run.func1+0x4d	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:103
0x44379e	sync.runtime_notifyListWait+0x12e														/usr/local/go/src/runtime/sema.go:267
0x4dbaed	sync.(*Cond).Wait+0x8d																/usr/local/go/src/sync/cond.go:57
0xa1c60b	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop+0x10b									/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:438
0xa187e0	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).processLoop+0x70								/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:147
0xa27a21	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).(k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.processLoop)-fm+0x41	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:121
0xf28a2f	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1+0x6f								/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:96
0xf280ed	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil+0xbd									/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97
0xf2800a	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until+0x5a										/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52
0xa1854f	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).Run+0x31f									/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/controller.go:121
0xa22f02	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).watchHandler+0x16e2	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:350
0xa20b5b	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch+0xe3b	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:321
0xa2817a	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).RunUntil.func1+0x4a	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:200
0xf28a2f	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1+0x6f	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:96
0xf280ed	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil+0xbd		/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97
0xf2800a	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until+0x5a			/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52
0xa2869d	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch.func1+0x42d	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflector.go:268
0xa5927a	k8s.io/kubernetes/federation/pkg/federation-controller/util.(*DelayingDeliverer).run+0x63a	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/util/delaying_deliverer.go:122
0xa5edf7	k8s.io/kubernetes/federation/pkg/federation-controller/util.(*DelayingDeliverer).StartWithHandler.func1+0x157	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/util/delaying_deliverer.go:174
0x49861d	testing.(*T).Run+0x56d		/usr/local/go/src/testing/testing.go:647
0x49cf09	testing.RunTests.func1+0xb9	/usr/local/go/src/testing/testing.go:793
0x498039	testing.tRunner+0xc9		/usr/local/go/src/testing/testing.go:610
0x4999fa	testing.RunTests+0x4ba		/usr/local/go/src/testing/testing.go:799
0x498bcf	testing.(*M).Run+0x12f		/usr/local/go/src/testing/testing.go:743
0x401428	main.main+0x1b8			k8s.io/kubernetes/federation/pkg/federation-controller/configmap/_test/_testmain.go:54
0x43122c	runtime.main+0x18c		/usr/local/go/src/runtime/proc.go:183
0x4a96bd	k8s.io/kubernetes/federation/pkg/federation-controller/configmap.(*ConfigMapController).Run.func1+0x4d	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/configmap/configmap_controller.go:238
0x5e2c02	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch.(*Broadcaster).loop+0x92	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch/mux.go:201
0xa309d5	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartEventWatcher.func1+0xc5	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:228
0xf22f1d	k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).flushDaemon+0x9d	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:879
0xa5ec86	k8s.io/kubernetes/federation/pkg/federation-controller/util.StartBackoffGC.func1+0x126	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/util/backoff.go:28
0x55725d	runtime/pprof.writeRuntimeProfile+0xbd								/usr/local/go/src/runtime/pprof/pprof.go:614
0x556ff9	runtime/pprof.writeGoroutine+0xc9								/usr/local/go/src/runtime/pprof/pprof.go:576
0x55287d	runtime/pprof.(*Profile).WriteTo+0x47d								/usr/local/go/src/runtime/pprof/pprof.go:298
0xf2ec46	k8s.io/kubernetes/federation/pkg/federation-controller/util/test.GetObjectFromChan+0x1b6	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/util/test/test_helper.go:248
0x4a7e3c	k8s.io/kubernetes/federation/pkg/federation-controller/configmap.GetConfigMapFromChan+0x3c	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/configmap/configmap_controller_test.go:164
0x4a758a	k8s.io/kubernetes/federation/pkg/federation-controller/configmap.TestConfigMapController+0x1d8a	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/configmap/configmap_controller_test.go:146
0x498039	testing.tRunner+0xc9										/usr/local/go/src/testing/testing.go:610
<autogenerated>:10: 
                        
	Error Trace:	configmap_controller_test.go:147
	
	Error:		Expected value not to be nil.
	
panic: runtime error: invalid memory address or nil pointer dereference
/usr/local/go/src/runtime/panic.go:500 +0x1ae
/usr/local/go/src/testing/testing.go:579 +0x474
/usr/local/go/src/runtime/panic.go:458 +0x271
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/federation/pkg/federation-controller/configmap/configmap_controller_test.go:148 +0x1e2e
/usr/local/go/src/testing/testing.go:610 +0xca
```
Seems like a different failure mode
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/42480/pull-kubernetes-unit/21387/

--- a/federation/pkg/federation-controller/configmap/configmap_controller.go
+++ b/federation/pkg/federation-controller/configmap/configmap_controller.go
@@ -323,13 +323,18 @@ func (configmapcontroller *ConfigMapController) reconcileConfigMap(configmap typ
 		glog.V(8).Infof("Skipping not federated config map: %s", key)
 		return
 	}
-	baseConfigMap := baseConfigMapObj.(*apiv1.ConfigMap)
+	obj, err := api.Scheme.DeepCopy(baseConfigMapObj)
+	configMap, ok := obj.(*apiv1.ConfigMap)
+	if err != nil || !ok {
+		glog.Errorf("Error in retrieving obj from store: %v, %v", ok, err)
+		return
+	}
 
 	// Check if deletion has been requested.
-	if baseConfigMap.DeletionTimestamp != nil {
-		if err := configmapcontroller.delete(baseConfigMap); err != nil {
+	if configMap.DeletionTimestamp != nil {
+		if err := configmapcontroller.delete(configMap); err != nil {
 			glog.Errorf("Failed to delete %s: %v", configmap, err)
-			configmapcontroller.eventRecorder.Eventf(baseConfigMap, api.EventTypeNormal, "DeleteFailed",
+			configmapcontroller.eventRecorder.Eventf(configMap, api.EventTypeNormal, "DeleteFailed",
 				"ConfigMap delete failed: %v", err)
 			configmapcontroller.deliverConfigMap(configmap, 0, true)
 		}
@@ -337,18 +342,18 @@ func (configmapcontroller *ConfigMapController) reconcileConfigMap(configmap typ
 	}
 
 	glog.V(3).Infof("Ensuring delete object from underlying clusters finalizer for configmap: %s",
-		baseConfigMap.Name)
+		configMap.Name)
 	// Add the required finalizers before creating a configmap in underlying clusters.
-	updatedConfigMapObj, err := configmapcontroller.deletionHelper.EnsureFinalizers(baseConfigMap)
+	updatedConfigMapObj, err := configmapcontroller.deletionHelper.EnsureFinalizers(configMap)
 	if err != nil {
 		glog.Errorf("Failed to ensure delete object from underlying clusters finalizer in configmap %s: %v",
-			baseConfigMap.Name, err)
+			configMap.Name, err)
 		configmapcontroller.deliverConfigMap(configmap, 0, false)
 		return
 	}
-	baseConfigMap = updatedConfigMapObj.(*apiv1.ConfigMap)
+	configMap = updatedConfigMapObj.(*apiv1.ConfigMap)
 
-	glog.V(3).Infof("Syncing configmap %s in underlying clusters", baseConfigMap.Name)
+	glog.V(3).Infof("Syncing configmap %s in underlying clusters", configMap.Name)
 
 	clusters, err := configmapcontroller.configmapFederatedInformer.GetReadyClusters()
 	if err != nil {
@@ -368,12 +373,12 @@ func (configmapcontroller *ConfigMapController) reconcileConfigMap(configmap typ
 
 		// Do not modify data.
 		desiredConfigMap := &apiv1.ConfigMap{
-			ObjectMeta: util.DeepCopyRelevantObjectMeta(baseConfigMap.ObjectMeta),
-			Data:       baseConfigMap.Data,
+			ObjectMeta: util.DeepCopyRelevantObjectMeta(configMap.ObjectMeta),
+			Data:       configMap.Data,
 		}
 
 		if !found {
-			configmapcontroller.eventRecorder.Eventf(baseConfigMap, api.EventTypeNormal, "CreateInCluster",
+			configmapcontroller.eventRecorder.Eventf(configMap, api.EventTypeNormal, "CreateInCluster",
 				"Creating configmap in cluster %s", cluster.Name)
 
 			operations = append(operations, util.FederatedOperation{
@@ -386,7 +391,7 @@ func (configmapcontroller *ConfigMapController) reconcileConfigMap(configmap typ
 
 			// Update existing configmap, if needed.
 			if !util.ConfigMapEquivalent(desiredConfigMap, clusterConfigMap) {
-				configmapcontroller.eventRecorder.Eventf(baseConfigMap, api.EventTypeNormal, "UpdateInCluster",
+				configmapcontroller.eventRecorder.Eventf(configMap, api.EventTypeNormal, "UpdateInCluster",
 					"Updating configmap in cluster %s", cluster.Name)
 				operations = append(operations, util.FederatedOperation{
 					Type:        util.OperationTypeUpdate,
@@ -404,7 +409,7 @@ func (configmapcontroller *ConfigMapController) reconcileConfigMap(configmap typ
 	}
 	err = configmapcontroller.federatedUpdater.UpdateWithOnError(operations, configmapcontroller.updateTimeout,
 		func(op util.FederatedOperation, operror error) {
-			configmapcontroller.eventRecorder.Eventf(baseConfigMap, api.EventTypeNormal, "UpdateInClusterFailed",
+			configmapcontroller.eventRecorder.Eventf(configMap, api.EventTypeNormal, "UpdateInClusterFailed",
 				"ConfigMap update in cluster %s failed: %v", op.ClusterName, operror)
 		})
 

--- a/federation/pkg/federation-controller/configmap/configmap_controller_test.go
+++ b/federation/pkg/federation-controller/configmap/configmap_controller_test.go
@@ -143,11 +143,18 @@ func TestConfigMapController(t *testing.T) {
 	}
 
 	configmapWatch.Modify(configmap1)
-	updatedConfigMap = GetConfigMapFromChan(cluster1UpdateChan)
-	assert.NotNil(t, updatedConfigMap)
-	assert.Equal(t, configmap1.Name, updatedConfigMap.Name)
-	assert.Equal(t, configmap1.Namespace, updatedConfigMap.Namespace)
-	assert.True(t, util.ConfigMapEquivalent(configmap1, updatedConfigMap))
+	for {
+		updatedConfigMap := GetConfigMapFromChan(cluster1UpdateChan)
+		assert.NotNil(t, updatedConfigMap)
+		if updatedConfigMap == nil {
+			break
+		}
+		assert.Equal(t, configmap1.Name, updatedConfigMap.Name)
+		assert.Equal(t, configmap1.Namespace, updatedConfigMap.Namespace)
+		if util.ConfigMapEquivalent(configmap1, updatedConfigMap) {
+			break
+		}
+	}
 
 	// Test add cluster
 	clusterWatch.Add(cluster2)

